crates/db_schema/src/impls/federation_blocklist.rs

```diff
@@ -1,27 +1,67 @@
 use crate::{
-  schema::federation_blocklist,
+  newtypes::PersonId,
+  schema::{admin_block_instance, federation_blocklist, instance},
   source::{
-    federation_blocklist::{FederationBlockList, FederationBlockListForm},
+    federation_blocklist::{BlockInstanceAction, FederationBlockList, FederationBlockListForm},
     instance::Instance,
+    moderator::{AdminBlockInstance, AdminBlockInstanceForm},
   },
+  traits::Crud,
   utils::{get_conn, DbPool},
 };
-use diesel::{dsl::insert_into, result::Error};
+use diesel::{
+  dsl::insert_into,
+  result::Error,
+  BoolExpressionMethods,
+  ExpressionMethods,
+  JoinOnDsl,
+  QueryDsl,
+};
 use diesel_async::{AsyncPgConnection, RunQueryDsl};
+use std::collections::{HashMap, HashSet};
 
 impl FederationBlockList {
-  pub async fn replace(pool: &DbPool, list_opt: Option<Vec<String>>) -> Result<(), Error> {
+  pub async fn replace(
+    pool: &DbPool,
+    list_opt: Option<Vec<BlockInstanceAction>>,
+    admin_person_id: PersonId,
+  ) -> Result<(), Error> {
     let conn = &mut get_conn(pool).await?;
     conn
       .build_transaction()
       .run(|conn| {
         Box::pin(async move {
           if let Some(list) = list_opt {
+            let existing_blocked_instances: HashMap<String, Option<String>> =
+              federation_blocklist::table
+                .inner_join(instance::table)
+                .inner_join(
+                  admin_block_instance::table.on(
+                    instance::id
+                      .eq(admin_block_instance::instance_id)
+                      .and(admin_block_instance::blocked),
+                  ),
+                )
+                .order_by((instance::id, admin_block_instance::when_.desc()))
+                .distinct_on(instance::id)
+                .select((instance::domain, admin_block_instance::reason))
+                .get_results::<(String, Option<String>)>(conn)
+                .await?
+                .into_iter()
+                .collect::<HashMap<String, Option<String>>>();
+
             Self::clear(conn).await?;
 
-            for domain in list {
+            let mut added_or_updated_blocks = Vec::new();
+            let mut unblocked_instances = existing_blocked_instances
+              .keys()
+              .cloned()
+              .collect::<HashSet<String>>();
```

Comment:
Really weird joining, none of it is necessary. You only need to add the AdminBlockInstance view to the moderation tables.

Reply:
i think this is necessary unless i'm misunderstanding.

the updates to the blocklist come as a complete set of blocked domains + reasons.
we've then got existing blocked domains + reasons and need to figure out what to put in the mod table, but there can be:
- newly added domains;
- existing domains with new reasons;
- existing unchanged domains;
- removed domains.

presumably we want to add entries for any newly added domains or existing domains with new reasons with a `blocked` flag since they represent admin actions. we also want to add entries for removed domains with the `blocked` flag false to represent the admin removing that domain.

i think the removed part is definitely necessary, the part figuring out created/edited could be removed, but then if you've got a list of 100 blocked domains and the admin edits the reason for one of them, or just removes one of them you'd be putting 100 unnecessary items in the admin log